src/array.jl

```diff
@@ -71,9 +71,13 @@ mutable struct CuArray{T,N,M} <: AbstractGPUArray{T,N}
     else
       maxsize
     end
-    data = DataRef(pool_free, pool_alloc(M, bufsize))
-    obj = new{T,N,M}(data, maxsize, 0, dims)
-    finalizer(unsafe_free!, obj)
+
+    GPUArrays.cached_alloc((CuArray, CUDA.device(), T, bufsize, M)) do
+        data = DataRef(pool_free, pool_alloc(M, bufsize))
+        obj = new{T,N,M}(data, maxsize, 0, dims)
+        finalizer(unsafe_free!, obj)
+        return obj
+    end::CuArray{T, N, M}
```

Comment:
Can we further improve cachability by caching the allocated, untyped buffer, and not the resulting array?

Reply:
Oh, that's too bad: You rely on `GPUArrays.storage`, effectively requiring the cached objects to be of type AbstractGPUArray...

Reply:
Back then I didn't want to deal with array finalizers as we'd need to check if the underlying buffer is cached (e.g. array is no longer held by anything so it is GC'd, but its buffer was cached, so we need to avoid freeing it).